### ISSUE TITLE
How to apply your method for selecting from my own data pool?

### ISSUE THREAD

**Issue by HelloWorld4747** [hamishivi/automated-instruction-selection#35]
First things first, I truly appreciate you sharing your exceptional work. Thanks!

My goal is to apply your method to select from my own data source. My data source is composed of a random number of samples collected from five datasets: Stanford Alpaca, Flan_v2, Open Assistant 1, WizardLM, and Dolly. I mixed them up into one dataset. The datasets used for evaluation in my case are completely different: MMLU, TruthfulQA, GSM8K, BBH, and TydiQA. So, for my case, **what should be my eval dataset ?** Referring to your example (the first step for creating an index for the train and eval data), does this mean I should use one of the datasets used for my evaluation as well? But, do I need to consider any criteria, or can I just randomly select one dataset as my eval dataset? Additionally, were the results presented in both Table 1 and Table 2 obtained by evaluating on the same datasets employed in your data selection pipeline?

Looking forward to your response soon!
Cheers!

**Comment by HelloWorld4747** [comment]
Also, I just realised that:
```
Step 1:
python -m minimal_multitask.compute_influence_cosinesim \
    --model_name meta-llama/Llama-2-7b-hf \
    --seed 42 \
    --train_dataset data/training_data/tulu_v2_unfiltered_data_dedup.jsonl \
    --eval_dataset alpacafarm \
    --index_path rds_200k_exp_from_base_weighted_mean_pool/cosine_train_reps.pt \
    --save_dir rds_200k_exp_from_base_weighted_mean_pool/ \
    --batch_size 1 \
    --pooling weighted_mean
```
```
Step 2:
for dataset in alpacafarm gsm8k_shots bbh_shots tydiqa_shots codex squad mmlu_shots; do
    python -m minimal_multitask.compute_influence_cosinesim \
        --model_name meta-llama/Llama-2-7b-hf \
        --seed 42 \
        --train_dataset data/training_data/tulu_v2_unfiltered_data_dedup.jsonl \
        --eval_dataset $dataset \
        --index_path rds_200k_exp_from_base_weighted_mean_pool/cosine_train_reps.pt \
        --save_dir rds_200k_exp_from_base_weighted_mean_pool/ \
        --batch_size 1 \
        --pooling weighted_mean
done
```
Are these two code snippets the same (for the step 1, you pass the alpacafarm to the --eval_dataset; for the step 2, it seems like just repeat the step 1 for all the eval dataset and the dataset alpacafarm is once again passed to --eval_dataset.)? Then, what is the point of doing step 1 solely?

Please do correct me if any of my understanding is incorrect.
Thanks!

**Comment by hamishivi** [comment]
Hiya, thanks for your interest!

1. In the two code snippets, the first is used to calculate the embeddings for the overall *train* pool, which we save to the given`index_path`. Once this is done, we run the second step, which loads the embeddings for the train pool from that path, and computes the embeddings and scores for the given dev set (alpacafarm gsm8k_shots bbh_shots tydiqa_shots codex squad mmlu_shots). The script used is the same (basically, it computes the train embeddings if it doesn't find anything at `index_path`), which is why it looks the same. You could just run the second command only, and know that on the first loop, it'll take much longer (since it takes a while to process train embeddings).

2. For evaluation dataset, MMLU, GSM8K, BBH, and TydiQA are all implemented, and you want to use `mmlu_shots, gsm8k_shots, bbh_shots tydiqa_shots`. These use the few-shot data in the task as query points to select data. For TruthfulQA, you need to subclass `TestDataset` and write an implementation in https://github.com/hamishivi/automated-instruction-selection/blob/main/minimal_multitask/data.py. You can look at the other datasets implemented there for help: basically, you need to load the data, and then format it to match the tulu setup. These prompts are used as query points for selection (so we embed them, and then use cosine similarity with them to rank training points). Then, for running evaluations, you can use the implementations in this repository, or whatever other evaluation implementation you want on the trained model.

Hope that makes sense and helps!

**Comment by HelloWorld4747** [comment]
> Hiya, thanks for your interest!
> 
> 1. In the two code snippets, the first is used to calculate the embeddings for the overall _train_ pool, which we save to the given`index_path`. Once this is done, we run the second step, which loads the embeddings for the train pool from that path, and computes the embeddings and scores for the given dev set (alpacafarm gsm8k_shots bbh_shots tydiqa_shots codex squad mmlu_shots). The script used is the same (basically, it computes the train embeddings if it doesn't find anything at `index_path`), which is why it looks the same. You could just run the second command only, and know that on the first loop, it'll take much longer (since it takes a while to process train embeddings).
> 2. For evaluation dataset, MMLU, GSM8K, BBH, and TydiQA are all implemented, and you want to use `mmlu_shots, gsm8k_shots, bbh_shots tydiqa_shots`. These use the few-shot data in the task as query points to select data. For TruthfulQA, you need to subclass `TestDataset` and write an implementation in https://github.com/hamishivi/automated-instruction-selection/blob/main/minimal_multitask/data.py. You can look at the other datasets implemented there for help: basically, you need to load the data, and then format it to match the tulu setup. These prompts are used as query points for selection (so we embed them, and then use cosine similarity with them to rank training points). Then, for running evaluations, you can use the implementations in this repository, or whatever other evaluation implementation you want on the trained model.
> 
> Hope that makes sense and helps!

Thank you so much for your clarification!

Assuming I have already done the implementation of aligning TruthfulQA to the Tulu setup, do you mean:
My running command should be:
```
Step 1 & 2:
for dataset in mmlu_shots gsm8k_shots bbh_shots tydiqa_shots truthfulqa_shots; do
    python -m minimal_multitask.compute_influence_cosinesim \
        --model_name meta-llama/Llama-2-7b-hf \
        --seed 42 \
        --train_dataset /path/to/my/own/dataset.jsonl \
        --eval_dataset $dataset \
        --index_path rds_200k_exp_from_base_weighted_mean_pool/cosine_train_reps.pt \
        --save_dir rds_200k_exp_from_base_weighted_mean_pool/ \
        --batch_size 1 \
        --pooling weighted_mean
done
```
Is my understanding correct?

Also, I wonder if the data sample used for the evaluation stage (this **evaluation stage** means the actual test sets used to assess the fine-tuned LLM) is the same as that used for the data selection (here I refer to the dataset passed to the **--eval_dataset**). The main thing is that I’m not sure whether there’s an issue of data contamination, or maybe I’m just overthinking it?

Besides, for the notation mentioned in your paper, *V* refers to the query set, and *D* stands for the source data pool, right? I wonder what a query set actually is. In practice, we have our own data source, which is *D*, and the corpus selected from *D* is represented as *d* in your paper. So, what is *V* in our case?

Thanks again! Really appreciate your effort on this work!

**Comment by hamishivi** [comment]
> My running command should be: 

Yup! that looks good to me. You'll just need to make sure `truthfulqa_shots` is in the mapping in that data file.

> Also, I wonder if the data sample used for the evaluation stage (this evaluation stage means the actual test sets used to assess the fine-tuned LLM) is the same as that used for the data selection

Good question! It's specifically not. We use samples from appropriate dev splits: for tasks with few-shot prompts, I use the few-shot examples (since we assume we have those labels) as the query set, and for tasks without these, I made my own test/dev splits. See Appendix A in https://arxiv.org/abs/2503.01807 for the details.

> Besides, for the notation mentioned in your paper, V refers to the query set, and D stands for the source data pool, right?

Yup, thats right! *V* is basically the data that is representative of the type of data we want to select for. If you have specific tasks you know  you want to optimize for, then it is the appropriate dev set for that. Alternatively, it could also be generic data that represents some high-quality model responses you want to select data for (for example, in the paper we explore using WildChat data or Arena-hard data as V/query set instead of task-specific dev sets, and find Arena-hard does pretty well as a generic query set).

**Comment by HelloWorld4747** [comment]
> > My running command should be:
> 
> Yup! that looks good to me. You'll just need to make sure `truthfulqa_shots` is in the mapping in that data file.
> 
> > Also, I wonder if the data sample used for the evaluation stage (this evaluation stage means the actual test sets used to assess the fine-tuned LLM) is the same as that used for the data selection
> 
> Good question! It's specifically not. We use samples from appropriate dev splits: for tasks with few-shot prompts, I use the few-shot examples (since we assume we have those labels) as the query set, and for tasks without these, I made my own test/dev splits. See Appendix A in https://arxiv.org/abs/2503.01807 for the details.
> 
> > Besides, for the notation mentioned in your paper, V refers to the query set, and D stands for the source data pool, right?
> 
> Yup, thats right! _V_ is basically the data that is representative of the type of data we want to select for. If you have specific tasks you know you want to optimize for, then it is the appropriate dev set for that. Alternatively, it could also be generic data that represents some high-quality model responses you want to select data for (for example, in the paper we explore using WildChat data or Arena-hard data as V/query set instead of task-specific dev sets, and find Arena-hard does pretty well as a generic query set).

Awesome! Thank you so much for your answer! 

For your key finding, *using Arena-hard data as V/query set offers a better option*, could you please point out where the corresponding code snippet I should refer to?

**Comment by hamishivi** [comment]
Yup!

The relevant code is here: https://github.com/hamishivi/automated-instruction-selection/blob/main/minimal_multitask/data.py#L1373

You can just specify "arena_hard" as `eval_dataset`.  You'll need to download the eval data (./shell_scripts/download_eval_data.sh) for this, since it depends on the arena hard jsonl in that zip.

**Comment by HelloWorld4747** [comment]
> Yup!
> 
> The relevant code is here: https://github.com/hamishivi/automated-instruction-selection/blob/main/minimal_multitask/data.py#L1373
> 
> You can just specify "arena_hard" as `eval_dataset`. You'll need to download the eval data (./shell_scripts/download_eval_data.sh) for this, since it depends on the arena hard jsonl in that zip.

Thank you so much for your clarification! It makes much more sense now.

**Comment by HelloWorld4747** [comment]
> Yup!
> 
> The relevant code is here: https://github.com/hamishivi/automated-instruction-selection/blob/main/minimal_multitask/data.py#L1373
> 
> You can just specify "arena_hard" as `eval_dataset`. You'll need to download the eval data (./shell_scripts/download_eval_data.sh) for this, since it depends on the arena hard jsonl in that zip.

Sorry for bothering you again!

Another question: **Can I use only the arena_hard dataset as a substitute for the other datasets when performing Step 1 & 2 (index creation), in order to save time?** My intuition is that arena_hard provides capabilities equivalent to those of the other datasets. Is this understanding correct?

**Comment by hamishivi** [comment]
Arena hard should do well for some tasks, but I can't guarantee it being universally great. It did provide strong broad capabilities when tested in the paper, though!

**Comment by HelloWorld4747** [comment]
> Arena hard should do well for some tasks, but I can't guarantee it being universally great. It did provide strong broad capabilities when tested in the paper, though!

Ok! Thank you so much for your clarification! ;-)
